Add PythonAPIIntegration class, context properties and macros.

Niloth-p opened the pull request:
### Sequence of Commits

1. Removed an unused extension variable from WebhookIntegrations. (This is required for the next commit).
2. Add DEFAULT_DOC_PATH to Integration.
3. Add PythonAPIIntegration class (integrations that exist in zulip/python-zulip-api/zulip/integrations), and split off those integrations' definitions.
4. Add a `directory_name` field for the PythonAPIIntegrations.
5. Add the new context properties (for the PythonAPIIntegrations, using the `directory_name`).
6. Add the new macros (for the PythonAPIIntegrations, using the added context properties).

---

Adds new context properties and macro(s) for the integrations located in python-zulip-api repo.

New context properties:
- `config_file_path`
- `integration_path`

#### Usage of the new context properties

I have not updated the integ docs to use these new additions yet, that would be done as part of updating each doc, in its own PR. But, here are some examples where this would be useful:
- Codebase, Git, OpenShift, Perforce, Subversion, Trac, Trello integrations - all have configuration files that need to be edited.
- About a dozen integrations have a non-empty requirements.txt file and would use the install-requirements macro that uses the `integration_path`. We could technically add a context property `requirements_path` too, but if we're using a macro, then that should be the only place where that context property would ever be used, so I didn't add that to the context.
- There are a few integrations that require copying some scripts from the `integration_path` and into the external plugin/service directory. So, we will be using the `integration_path` in their docs, for the copy-instruction.

#### Macros

There are a few more macros that we could consider adding, based on common patterns, after discussing on CZO.
The macros are included in this PR for now as the macro(s) are to be built upon the context properties.

#### Design Decision

~~The introduction of context properties and macro(s) may clash with the plan of moving the corresponding docs to the python-zulip-api repo.~~

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Niloth-p commented on zerver/views/documentation.py:
```diff
@@ -389,6 +389,13 @@ def integration_doc(request: HttpRequest, *, integration_name: PathOnly[str]) ->
+        context["config_file_path"] = (
+            f"/usr/local/share/zulip/integrations/{integration.name}/zulip_{integration.name}_config.py"
+        )
+        context["integration_path"] = f"/usr/local/share/zulip/integrations/{integration.name}"
```
The two values are not 100% accurate yet. 
- We would need to rename these integration directories in python-zulip-api. I've left this update for later. When we start using the parameters in each doc, we could update the python-zulip-api repo in parallel. The update would be needed in most cases, e.g., for renaming "stream" -> "channel" in the config files.
	- irc bridge -> irc
	- matrix bridge -> matrix
	- hg -> mercurial
	- google -> google-calendar
- For integrations that don't use a config file, the two context properties wouldn't apply, but we wouldn't use the macro for those docs either, so I assumed it should be fine.
- For cases where the config file is not in python (example: `matrix_bridge.conf`), we could either pass those `config_file_path` values through their declarations in `Integration`, or we could use the `integration_path` and append the config file name. Since the number of integrations that need this are very few, I think we can just go with the latter option.

timabbott replied:
I would prefer to not need to reorganize the `python-zulip-api` repository as part of this PR; we should probably just have a parameter used in declaring the relevant integration subclass for mapping the integration name to the directory name.

Niloth-p reviewed (commented):
Self-reviewed the addition of the new context properties: `config_file_path` and `integration_path`

timabbott commented on zerver/views/documentation.py:
```diff
@@ -389,6 +389,13 @@ def integration_doc(request: HttpRequest, *, integration_name: PathOnly[str]) ->
             context["all_event_types"] = all_event_types
     if isinstance(integration, HubotIntegration):
         context["hubot_docs_url"] = integration.hubot_docs_url
+    if not isinstance(integration, WebhookIntegration) and not isinstance(
```
I think you want it to be a positive `isinstance` check for a specific set of classes that should have this treatment, not a list of classes that should not have it.

timabbott commented on templates/zerver/integrations/include/install-requirements.md:
```diff
@@ -0,0 +1,3 @@
+Install the python packages required for this integration by running:
+
+`pip install -r {{ integration_path }}/requirements.txt`
```
This seems to not be used yet?

timabbott commented:
Looks good to me code-wise, please resolve the merge conflicts and do `diff` testing to confirm no unexpected user-facing changes.

Niloth-p commented:
**Updates**

- Resolved merge conflicts - these were deep, because most entries in the list had been updated with their logo parameters removed.
- Performed diff testing of the `/integrations` page.
- Performed print verification for `integration.doc`, though it should be covered by the `test-docs`.
- Added the 2nd commit - it just renames the `bigbluebutton` image files to `big-blue-button`, matching the integration name.
- Removed the last commit adding a new macro. Regarding the question above, it did have a commit description with "This macro is planned to be used in the docs of all PythonAPIIntegrations that have a non-empty requirements.txt file." But, I've removed this commit for the moment, because we might want to consider using the `--provision` option instead of asking the users to run the `pip install -r requirements.txt`. The integrations would need to be updated for `--provision` to work, but that might be a better option. Leaving that out for a separate PR.
